### PR TITLE
Add Ellis' code to skeleton

### PR DESCRIPTION
**Pull request by EllisLempriere**

### CODE REVIEW

File: src/main/java/StringManipulation.java [hunk 1]
```diff
@@ -1,33 +1,134 @@
...
 
     @Override
     public String removeNthCharacter(int n, boolean maintainSpacing) {
-        return null;
+        if (n <= 0)
+            throw new IllegalArgumentException();
+        if (n > string.length())
+            throw new IndexOutOfBoundsException();
+
+        StringBuilder newStr = new StringBuilder();
```
Comment (WNeil) [inline src/main/java/StringManipulation.java:40]: StringBuilder in retrospect feels a lot simpler than the solution I came up with, atleast in this case haha

File: src/test/java/StringManipulationTest.java [hunk 2]
```diff
@@ -1,149 +1,370 @@
 import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.DisplayName;
```
Comment (WNeil) [inline src/test/java/StringManipulationTest.java:3]: Good little organizational choice, definitely taking this in the future lol.

File: src/main/java/StringManipulation.java [hunk 3]
```diff
@@ -1,33 +1,134 @@
...
+                // Case where end of word is found
+            } else if (!isSpaceChar(string.charAt(i)) && isSpaceChar(string.charAt(i + 1))) {
+                if (wordCount >= startWord && wordCount <= endWord) {
+                    sb.append(string.charAt(i));
+                    words[wordIndex++] = sb.toString();
+                    sb = new StringBuilder();
+                }
+
+            } else if (!isSpaceChar(string.charAt(i)) && wordCount >= startWord && wordCount <= endWord) {
+                sb.append(string.charAt(i));
```
Comment (WNeil) [inline src/main/java/StringManipulation.java:87]: StringBuilder is a good choice, I however favored the split method of the string class  using the regex \\s+, this gives you an array of all the char groups between whitespace gaps, then you can just loop through that array and add desired words to your final words array. that solution is a bit shorter to write and simpler IMHO

File: src/main/java/StringManipulation.java [hunk 4]
```diff
@@ -1,33 +1,134 @@
...
+
+            chars[indices[i]] = string.charAt(i);
+            indexUsed[indices[i]] = true;
+        }
+
+        StringBuilder sb = new StringBuilder();
+        for (char c : chars)
+            sb.append(c);
+
+        return sb.toString();
```
Comment (WNeil) [inline src/main/java/StringManipulation.java:119]: Wow wow wow, I really like the boolean array decision, very smart idea to catch something I did not even account for! I think you can construct a new String from a char array without needing to pass it through a StringBuilder, making a new String object with the char array as a param, otherwise really good job!

File: src/test/java/StringManipulationTest.java [hunk 5]
```diff
@@ -1,149 +1,370 @@
...
-        fail("Not yet implemented");
+        // arrange
+        manipulatedString.setString("Here are spaces    ");
+        int expected = 3;
+
+        // act
+        int actual = manipulatedString.count();
+
+        // assert
+        assertEquals(expected, actual);
```
Comment (WNeil) [inline src/test/java/StringManipulationTest.java:105]: Really small detail I noticed in your written tests is the declaration of a separate int over just using the number in the parameter, I assume this is an organizational choice due to the comments, weirded me out at first but I can definitely appreciate the tidiness after reading it haha

File: src/main/java/StringManipulation.java [hunk 6]
```diff
@@ -1,33 +1,134 @@
...
+
+        int count = 0;
+        for (int i = 0; i < string.length() - 2; i++) {
+            if (i == 0 && !isSpaceChar(string.charAt(i)))
+                count++;
+            else if (isSpaceChar(string.charAt(i)) && !isSpaceChar(string.charAt(i + 1)))
+                count++;
+        }
+
+        return count;
```
Comment (WNeil) [inline src/main/java/StringManipulation.java:30]: Pretty laborious solution, I like the meticulousness and I see the hussle, but like I say in a later comment you can use the split method of the string class with the regex "\\s+", giving you a string array of everything split by whitespaces, and with that you could just return the length of that generated array, easy peasy.

**Review by WNeil — APPROVED** [review]
In Conclusion I think the implementation and test cases made are very much satisfactory. The choice of using StringBuilder in the StringManipulation class let this code be very meticulous in its strategy and as a result will probably catch more edge case weirdness than the solutions I constructed, though perhaps at the slight cost of overcomplication in some scenarios.
